sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala

```diff
@@ -33,10 +33,13 @@ case class ReuseAdaptiveSubquery(
 
     plan.transformAllExpressionsWithPruning(_.containsPattern(PLAN_EXPRESSION)) {
       case sub: ExecSubqueryExpression =>
-        val newPlan = reuseMap.getOrElseUpdate(sub.plan.canonicalized, sub.plan)
-        if (newPlan.ne(sub.plan)) {
-          sub.withNewPlan(ReusedSubqueryExec(newPlan))
-        } else {
+        // `InsertAdaptiveSparkPlan` compiles subquery for each exprId, then the java object
+        // is always `eq` if two subqueries have same exprId.
+        // Check if the subquery can be reused manually instead of call `getOrElseUpdate`.
+        reuseMap.get(sub.plan.canonicalized).map { subquery =>
+          sub.withNewPlan(ReusedSubqueryExec(subquery))
+        }.getOrElse {
+          reuseMap.put(sub.plan.canonicalized, sub.plan)
```

Comment:
do we have atomicity concerns?

Reply:
The behavior should be same with `getOrElseUpdate`. I changed `put` to `putIfAbsent` to fully equivalent with `getOrElseUpdate` that the first `put` will win if there is race condition.

Reply:
Is this correct? 2 concurent `.get()`s for 2 semantic equal plans can return `None`s and none of the plans will be wrapped into `ReusedSubqueryExec`.

Reply:
The behavior follows the `getOrElseUpdate`. It does not affect perf and I'm not sure how many possible can that happen, so I do not lock it.

Reply:
I think `getOrElseUpdate()` on `TrieMap` is atomic, a `.get()` followed by a `.put()` is not.

Reply:
I mean, `getOrElseUpdate()` is basically a `.get()` followed by a `.putIfAbsent()` as you do. But you should then check if the map contains the key you put into it and depending on that do the reuse.

Reply:
thank you @peter-toth , I see it now

Reply:
Thanks! https://github.com/apache/spark/pull/41046/commits/99f810c4084576eea19231590f4ddf9a9afaf7b4 looks ok to me.